src/workers/tx_sender.rs

```diff
@@ -60,7 +58,57 @@ impl TxSender {
         Self {
             tpu_manager,
             txs_sent: Default::default(),
-            counting_semaphore: Arc::new(tokio::sync::Semaphore::new(5)),
+        }
+    }
+
+    /// retry enqued_tx(s)
+    async fn forward_txs(
+        &self,
+        sigs_and_slots: Vec<(String, u64)>,
+        txs: Vec<WireTransaction>,
+        postgres: Option<PostgresMpscSend>,
+    ) {
+        assert_eq!(sigs_and_slots.len(), txs.len());
+
+        if sigs_and_slots.is_empty() {
+            return;
+        }
+
+        let tpu_client = self.tpu_manager.clone();
+        let txs_sent = self.txs_sent.clone();
+
+        let quic_response = match tpu_client.try_send_wire_transaction_batch(txs).await {
+            Ok(_) => {
+                for (sig, _) in &sigs_and_slots {
+                    txs_sent.insert(sig.to_owned(), TxProps::default());
+                }
+                // metrics
+                TXS_SENT.inc_by(sigs_and_slots.len() as f64);
+
+                1
+            }
+            Err(err) => {
+                warn!("{err}");
+                0
+            }
+        };
+
+        if let Some(postgres) = postgres {
+            let forwarded_slot = tpu_client.get_tpu_client().await.estimated_current_slot();
+
+            for (sig, recent_slot) in sigs_and_slots {
+                postgres
+                    .send(PostgresMsg::PostgresTx(PostgresTx {
```

Comment:
no matter how much you try to re-batch on the other end of this channel, queueing all the individual inserts as their own message is the issue.
please follow the tx-backup example and generate batch insert queries: 1 query, inserting as many data tuples as you have.

you can then proceed to revert all this batching logic from the client-side, it's unnecessarily complex. the guiding principle is: when you receive the data it is already batched:

```
        sigs_and_slots: Vec<(String, u64)>,
        txs: Vec<WireTransaction>,
```

Reply:
send_tx should be send_tsx and receive a Vec<PostgresTx>
update_tx should be update_tsx and receive a Vec<PostgresUpdateTx>, which should contain signatures.